README
Contents
========

This package contains the source for PdfMasher. To learn how to build it, refer to the "Build PdfMasher" section. Below is the description of the various subfolders:

- core: Contains the core logic code for PdfMasher. Written in Python.
- cocoa: UI code for the Cocoa toolkit. It's Objective-C code.
- qt: UI code for the Qt toolkit. It's written in Python and uses PyQt.
- images: Images used by the different UI codebases.
- debian_*: Skeleton files required to create a .deb package
- help: Help document, written for Sphinx.

There are also other sub-folder that comes from external repositories (automatically checked out
with as mercurial subrepos):

- hscommon: A collection of helpers used across HS applications.
- cocoalib: A collection of helpers used across Cocoa UI codebases of HS applications.
- qtlib: A collection of helpers used across Qt UI codebases of HS applications.

PdfMasher Dependencies
======================

Before being able to build PdfMasher, a few dependencies have to be installed:

General dependencies
--------------------

- Python 3.2 (http://www.python.org)
- pdfminer3k (http://hg.hardcoded.net/pdfminer3k)
- jobprogress 1.0.0 (http://hg.hardcoded.net/jobprogress)
- Sphinx 1.1.0 (http://sphinx.pocoo.org/)
- pytest 2.0.3, to run unit tests. (http://pytest.org/)
- Markdown 2.1.1 (http://www.freewisdom.org/projects/python-markdown)

OS X prerequisites
------------------

- XCode's Command Line Tools
- ObjP 1.2.1 (http://bitbucket.org/hsoft/objp)
- xibless 0.5.0 (https://bitbucket.org/hsoft/xibless)

Windows prerequisites
---------------------

- Visual Studio 2008 (Express is enough) is needed to build C extensions. (http://www.microsoft.com/Express/)
- PyQt 4.7.5 (http://www.riverbankcomputing.co.uk/news)
- cx_Freeze, if you want to build a exe. You don't need it if you just want to run PdfMasher. (http://cx-freeze.sourceforge.net/)
- Advanced Installer, if you want to build the installer file. (http://www.advancedinstaller.com/)

Linux prerequisites
-------------------

- PyQt 4.7.5 (http://www.riverbankcomputing.co.uk/news)

The easy way!
-------------

There's an easy way to install the majority of the prerequisites above, and it's pip
(http://www.pip-installer.org/) which has recently started to support Python 3. So install it and
then run::

    pip install -r requirements-[osx|win|lnx].txt

([osx|win|lnx] depends, of course, on your platform). Because Sphinx doesn't support Python 3 yet,
it's not in the requirements file and you'll have to install it manually. You might have to compile
PyObjC manually too (see gotchas below). Advanced Installer, having nothing to do with Python, is a
manual install. PyQt also needs to be installed manually: the package is on PyPI, but there's no
download available.

Prerequisite gotchas
--------------------

Correctly installing the prerequisites is tricky. Make sure you have at least the version number 
required for each prerequisite.

If you didn't use mercurial to download this source, you probably have an incomplete source folder!
External projects (hscommon, qtlib, cocoalib) need to be at the root of the PdfMasher project
folder. You'll have to download those separately. Or use mercurial, it's much easier.

As far as I can tell, you don't *have* to compile/install everything manually and you can normally
use `easy_install` to install python dependencies. However, be aware that compiling/installing
manually from the repositories of each project is what I personally do, so if you hit a snag 
somewhere, you might want to try the manual way.

Another one on OS X: I wouldn't use macports/fink/whatever. Whenever I tried using those, I always 
ended up with problems.

Whenever you have a problem, always double-check that you're running the correct python version. 
You'll probably have to tweak your $PATH.

Building PdfMasher
==================

First, make sure you meet the dependencies listed in the section above. Then you need to configure
your build with:

	python configure.py
	
If you want, you can specify a UI to use with the `--ui` option. So, if you want to build PdfMasher
with Qt on OS X, then you have to type `python configure.py --ui=qt`. You can also use the `--dev`
flag to indicate a dev build (it will build `pyplugin.plugin` in alias mode and use the "dev" config
in XCode).

Then, just build the thing and then run it with:

	python build.py
	python run.py

If you want to create ready-to-upload package, run:

	python package.py


help/changelog
=== 0.7.4 (2013-03-30)

* Removed the fairware dialog ([More Info](http://www.hardcoded.net/articles/phasing-out-fairware)).
* Added a 64-bit build for Windows.

=== 0.7.3 (2013-01-27)

* Fixed bugs preventing some PDFs to be read.
* Minor fixes/improvements here and there.

=== 0.7.2 (2012-08-07)

* Fixed a (stupid) bug causing crashes on Save Project.

=== 0.7.1 (2012-07-20)

* Fixed bugs preventing some PDFs to be read.
* Fixed minor UI glitches. [Mac]

=== 0.7.0 (2012-07-10)

* Added the ability to save and load PdfMasher projects. (#15)
* Tweaked Fairware reminders.
* Improved Debian packaging. [Linux]
* Upgraded minimum requirements to Ubuntu 12.04.

=== 0.6.3 (2011-10-07)

* Fixed bugs preventing some PDFs to be read.

=== 0.6.2 (2011-09-05)

* Fixed bugs preventing some PDFs to be read.
* Fixed bug making TOC and footnotes in MOBI link to wrong locations.

=== 0.6.1 (2011-08-30)

* Fixed bugs preventing some PDFs to be read.
* Fixed a crash on clicking the page view.
* Fixed a crash on ebook generation.
* Fixed a bug causing the first character(s) of each line to be removed when reading a PDF.

=== 0.6.0 (2011-08-22)

* Added automatic page breaks for title elements (level 1 and 2).
* Added automatic TOC generation for title elements(level 1 and 2). (#4)
* Improved text layout detection in PDFs.
* Fixed bugs preventing some PDFs to be read.

=== 0.5.0 (2011-08-18)

* Added the ability to generate MOBI and EPUB documents (using code from Calibre).
* Changed the license from BSD to GPL (Calibre is GPL).
* Fixed a memory usage problem.

=== 0.4.0 (2011-08-09)

* Added a new Page View feature, which is really cool.
* Moved the Edit/Build tab view to the right to leave more space to the page view.
* Title elements now have multiple levels (1 to 6)
* Added keyboard bindings to quickly set element states.
* Improved text layout detection in PDFs.
* Fixed bugs preventing some PDFs to be read.

=== 0.3.0 (2011-07-26)

* Added a new flag: "To Fix".
* Added keyboard shortcuts for quickly flagging elements.
* Improved text layout detection in PDFs.
* Fixed bugs preventing some PDFs to be read.

=== 0.2.1 (2011-07-15)

* Fixed a bug where title elements wouldn't correctly be wrapped. (#3)
* Fixed a bug where generating markdown for elements with footnotes more than once would make a mess.
* Fixed bugs preventing some PDFs to be read.

=== 0.2.0 (2011-07-10)

* Made the PDF opening progress smooth (progress bar).
* Added a Markdown intermediate step in the build process.
* Added a Text Length column.
* Made the error message when opening a non-PDF file more friendly.
* Fixed a bug making the order of some text parts to be mixed up.
* Fixed bugs preventing some PDFs to be read.

=== 0.1.2 (2011-07-07)

* Fixed bugs preventing some PDFs to be read.
* Fixed a bug where there would sometimes be no space character between words.
* Added auto-update support with Sparkle. [Mac OS X]

=== 0.1.1 (2011-06-27)

* Fixed bugs preventing some PDFs to be read.
* Correctly set the encoding of resulting HTML files.

=== 0.1.0 (2011-06-21)

* Initial preview release.


build.py
#!/usr/bin/env python3
# Created By: Virgil Dupras
# Created On: 2011-06-18
# Copyright 2013 Hardcoded Software (http://www.hardcoded.net)
# 
# This software is licensed under the "GPL v3" License as described in the "LICENSE" file, 
# which should be included with this package. The terms are also available at 
# http://www.hardcoded.net/licenses/gplv3_license

import sys
import os
import os.path as op
import shutil
import json
from argparse import ArgumentParser
import compileall

from hscommon import sphinxgen
from hscommon.build import (print_and_do, copy_packages, get_module_version, filereplace,
    add_to_pythonpath, copy, copy_sysconfig_files_for_embed, OSXAppStructure, build_cocoa_ext,
    build_cocoalib_xibless, copy_all, copy_embeddable_python_dylib, collect_stdlib_dependencies)
from hscommon.plat import ISOSX
from hscommon.util import ensure_folder, delete_files_with_pattern

def parse_args():
    parser = ArgumentParser()
    parser.add_argument('--clean', action='store_true', dest='clean',
        help="Clean build folder before building")
    parser.add_argument('--cocoa-compile', action='store_true', dest='cocoa_compile',
        help="Build only Cocoa modules and executables")
    parser.add_argument('--xibless', action='store_true', dest='xibless',
        help="Build only xibless UIs")
    parser.add_argument('--doc', action='store_true', dest='doc',
        help="Build only the help file")
    args = parser.parse_args()
    return args

def build_xibless():
    import xibless
    ensure_folder('cocoa/autogen')
    xibless.generate('cocoa/ui/edit_pane.py', 'cocoa/autogen/PMEditPane_UI')
    xibless.generate('cocoa/ui/build_pane.py', 'cocoa/autogen/PMBuildPane_UI')
    xibless.generate('cocoa/ui/page_pane.py', 'cocoa/autogen/PMPageController_UI')
    xibless.generate('cocoa/ui/main_window.py', 'cocoa/autogen/PMMainWindow_UI')
    xibless.generate('cocoa/ui/main_menu.py', 'cocoa/autogen/PMMainMenu_UI')

def build_cocoa(dev):
    app = OSXAppStructure('build/PdfMasher.app')
    print('Generating Info.plist')
    app_version = get_module_version('core')
    filereplace('cocoa/InfoTemplate.plist', 'cocoa/Info.plist', version=app_version)    
    app.create('cocoa/Info.plist')
    print("Building the cocoa layer")
    build_cocoalib_xibless(withfairware=False)
    build_xibless()
    pydep_folder = op.join(app.resources, 'py')
    if not op.exists(pydep_folder):
        os.mkdir(pydep_folder)
    build_cocoa_proxy_module()
    build_cocoa_bridging_interfaces()
    copy_embeddable_python_dylib('build')
    tocopy = ['core', 'ebooks', 'hscommon', 'cocoa/inter', 'cocoalib/cocoa', 'jobprogress', 'objp',
        'cssutils', 'cssselect', 'pdfminer', 'lxml', 'ply', 'markdown', 'encutils']
    copy_packages(tocopy, pydep_folder, create_links=dev)
    copy('cocoa/pyplugin.py', 'build/pyplugin.py')
    sys.path.insert(0, 'build')
    collect_stdlib_dependencies('build/pyplugin.py', pydep_folder)
    del sys.path[0]
    # Views are not referenced by python code, so they're not found by the collector.
    copy_all('build/inter/*.so', op.join(pydep_folder, 'inter'))
    copy_sysconfig_files_for_embed(pydep_folder)
    if not dev:
        # Important: Don't ever run delete_files_with_pattern('*.py') on dev builds because you'll
        # be deleting all py files in symlinked folders.
        compileall.compile_dir(pydep_folder, force=True, legacy=True)
        delete_files_with_pattern(pydep_folder, '*.py')
        delete_files_with_pattern(pydep_folder, '__pycache__')
    os.chdir('cocoa')
    print("Compiling with WAF")
    os.system('{0} waf configure && {0} waf'.format(sys.executable))
    os.chdir('..')
    print("Creating the .app folder")
    app.copy_executable('cocoa/build/PdfMasher')
    resources = ['images/main_icon.icns', 'cocoa/dsa_pub.pem', 'build/pyplugin.py', 'build/help']
    app.copy_resources(*resources, use_symlinks=dev)
    app.copy_frameworks('build/Python', 'cocoalib/Sparkle.framework')
    print("Creating the run.py file")
    copy('cocoa/runtemplate.py', 'run.py')

def build_cocoa_proxy_module():
    print("Building Cocoa Proxy")
    import objp.p2o
    objp.p2o.generate_python_proxy_code('cocoalib/cocoa/CocoaProxy.h', 'build/CocoaProxy.m')
    build_cocoa_ext("CocoaProxy", 'cocoalib/cocoa',
        ['cocoalib/cocoa/CocoaProxy.m', 'build/CocoaProxy.m', 'build/ObjP.m',
            'cocoalib/HSErrorReportWindow.m', 'cocoa/autogen/HSErrorReportWindow_UI.m'],
        ['AppKit', 'CoreServices'],
        ['cocoalib', 'cocoa/autogen'])

def build_cocoa_bridging_interfaces():
    print("Building Cocoa Bridging Interfaces")
    import objp.o2p
    import objp.p2o
    add_to_pythonpath('cocoa')
    add_to_pythonpath('cocoalib')
    from cocoa.inter import (PyGUIObject, GUIObjectView, PyTable, TableView, PyColumns,
        ColumnsView, PyBaseApp, BaseAppView, PyTextField)
    from inter.app import PyPdfMasher, PdfMasherView
    from inter.build_pane import PyBuildPane
    from inter.edit_pane import PyEditPane, EditPaneView
    from inter.element_table import PyElementTable
    from inter.page_controller import PyPageController, PageControllerView
    from inter.page_repr import PyPageRepr, PageReprView
    allclasses = [PyGUIObject, PyTable, PyColumns, PyBaseApp, PyTextField, PyPdfMasher,
        PyBuildPane, PyEditPane, PyElementTable, PyPageController, PyPageRepr]
    for class_ in allclasses:
        objp.o2p.generate_objc_code(class_, 'cocoa/autogen', inherit=True)
    allclasses = [GUIObjectView, TableView, ColumnsView, BaseAppView, EditPaneView,
        PageControllerView, PageReprView, PdfMasherView]
    clsspecs = [objp.o2p.spec_from_python_class(class_) for class_ in allclasses]
    objp.p2o.generate_python_proxy_code_from_clsspec(clsspecs, 'build/CocoaViews.m')
    build_cocoa_ext('CocoaViews', 'cocoa/inter', ['build/CocoaViews.m', 'build/ObjP.m'])

def build_qt():
    print("Building resource file")
    qrc_path = op.join('qt', 'pm.qrc')
    pyrc_path = op.join('qt', 'pm_rc.py')
    print_and_do("pyrcc4 -py3 {0} > {1}".format(qrc_path, pyrc_path))
    print("Creating the run.py file")
    runtemplate_path = op.join('qt', 'runtemplate.py')
    shutil.copy(runtemplate_path, 'run.py')

def build_help():
    print("Generating Help")
    platform = 'osx' if ISOSX else 'win'
    current_path = op.abspath('.')
    confpath = op.join(current_path, 'help', 'conf.tmpl')
    help_basepath = op.join(current_path, 'help', 'en')
    help_destpath = op.join(current_path, 'build', 'help')
    changelog_path = op.join(current_path, 'help', 'changelog')
    tixurl = "http://bitbucket.org/hsoft/pdfmasher/issue/{0}"
    confrepl = {'platform': platform}
    sphinxgen.gen(help_basepath, help_destpath, changelog_path, tixurl, confrepl, confpath)

def build_normal(ui, dev):
    build_help()
    if ui == 'cocoa':
        build_cocoa(dev)
    elif ui == 'qt':
        build_qt()

def main():
    args = parse_args()
    conf = json.load(open('conf.json'))
    ui = conf['ui']
    dev = conf['dev']
    print("Building PdfMasher with UI {0}".format(ui))
    if dev:
        print("Building in Dev mode")
    if args.clean:
        if op.exists('build'):
            shutil.rmtree('build')
    if not op.exists('build'):
        os.mkdir('build')
    if args.doc:
        build_help()
    elif args.cocoa_compile:
        build_cocoa_proxy_module()
        build_cocoa_bridging_interfaces()
        os.chdir('cocoa')
        os.system('{0} waf configure && {0} waf'.format(sys.executable))
        os.chdir('..')
        copy('cocoa/build/PdfMasher', 'build/PdfMasher.app/Contents/MacOS/PdfMasher')
    elif args.xibless:
        build_xibless()
    else:
        build_normal(ui, dev)

if __name__ == '__main__':
    main()
